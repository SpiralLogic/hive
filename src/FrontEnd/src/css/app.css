body {
  --focus: #1712A2A0;
  --p0background: #FFEE0B;
  --p0creature: #000000;
  --p0outline0: #D2116A7d;
  --p1background: #000000;
  --p1creature: #FFEE0B;
  --p1outline: #8A2BE27d;
  --hex-offset: -1;
  --hex-size: 95px;
  --tile-size: 70px;
  margin: 0;
  overscroll-behavior-y: contain;
  height: 100vh;
  background: #4c4c4c;
}

body > svg {
  display: none;
}

@media only screen and (max-width: 720px) {
  body {
    filter: brightness(1.1);
    /*noinspection CssInvalidPropertyValue*/
    max-height: -webkit-fill-available;
    margin: 0;
    --hex-size: 60px;
    --tile-size: 45px;
  }
}
