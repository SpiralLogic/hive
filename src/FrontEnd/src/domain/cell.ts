import { HexCoordinates } from './hex-coordinates';
import { Tile } from './tile';

export type Cell = {
  coords: HexCoordinates;
  tiles: Tile[];
};
