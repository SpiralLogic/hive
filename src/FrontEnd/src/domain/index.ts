export * from './cell';
export * from './game-state';
export * from './move';
export * from './tile';
export * from './player';
export * from './cell';
export * from './Row';
export * from './hex-coordinates';
