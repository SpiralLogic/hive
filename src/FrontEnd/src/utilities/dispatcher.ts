import { useEffect } from 'preact/hooks';
import { HiveDispatcher, HiveEventListener, HiveIntent } from '../services';

const hiveDispatcher = new HiveDispatcher();

export const useHiveDispatcher = (): HiveDispatcher => hiveDispatcher;

export const addHiveDispatchListener = <T extends HiveIntent>(
  type: T['type'],
  listener: HiveEventListener<T>
): void => {
  useEffect(() => {
    const emitter = useHiveDispatcher();
    return emitter.add<T>(type, listener);
  });
};

export const dispatchHiveEvent = hiveDispatcher.dispatch;
