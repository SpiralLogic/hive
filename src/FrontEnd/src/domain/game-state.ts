import { Cell } from './cell';
import { Player } from './player';

export type GameState = {
  cells: Cell[];
  players: Player[];
  gameId: GameId;
  gameStatus: GameStatus
};

export type GameId = string;
export type GameStatus =
  | 'NewGame'
  | 'MoveSuccess'
  | 'AiWin'
  | 'Player0Win'
  | 'Player1Win'
  | 'MoveSuccessNextPlayerSkipped'
  | 'MoveInvalid'
  | 'GameOver' | 'Draw';
