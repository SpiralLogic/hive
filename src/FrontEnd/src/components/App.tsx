import { Fragment, FunctionComponent, h } from 'preact';
import { GameId, GameState, PlayerId } from '../domain';
import { MoveEvent, TileEvent } from '../utilities/hive-dispatcher';
import { addHiveEventListener, useHiveDispatcher } from '../utilities/hooks';
import { opponentSelectionHandler } from '../utilities/handlers';
import { useEffect, useState } from 'preact/hooks';
import Engine from '../utilities/game-engine';
import GameArea from './GameArea';
import Links from './Links';
import Rules from './Rules';

const App: FunctionComponent = () => {
  const [gameState, updateGameState] = useState<GameState | undefined>(undefined);
  const [playerId, setPlayerId] = useState<PlayerId>(0);
  const [showHelp, setShowHelp] = useState<boolean>(true);

  const initializeGame = (
    getInitial: (gameId: GameId) => Promise<GameState>,
    gameId: string,
    routePlayerId: string
  ) =>
    getInitial(gameId).then((gameState) => {
      const { gameId } = gameState;
      const [player] = gameState.players;
      const playerId = (Number(routePlayerId) as PlayerId) ? Number(routePlayerId) : player.id;
      updateGameState(gameState);
      setPlayerId(playerId);
      window.history.replaceState({ playerId, gameId }, document.title, `/game/${gameId}/${playerId}`);
    });

  useEffect(() => {
    const [, route, gameId, routePlayerId] = window.location.pathname.split('/');
    const loadExistingGame = gameId && route === 'game';
    const getInitial = loadExistingGame ? Engine.getExistingGame : Engine.getNewGame;
    initializeGame(getInitial, gameId, routePlayerId).then();
  }, []);

  if (gameState === undefined) return <h1>loading !</h1>;
  const hiveDispatcher = useHiveDispatcher();

  addHiveEventListener<MoveEvent>('move', async (event) => {
    const newGameState = await Engine.moveTile(gameState.gameId, event.move);
    updateGameState(newGameState);
  });

  useEffect(() => {
    const { closeConnection, sendSelection } = Engine.connectGame(gameState.gameId, {
      updateGameState,
      opponentSelection: opponentSelectionHandler,
    });

    const selectionChangeHandler = (event: TileEvent) => {
      if (sendSelection) {
        sendSelection('select', event.tile);
      }
    };

    const deselectionChangeHandler = (event: TileEvent) => {
      if (sendSelection) {
        sendSelection('deselect', event.tile);
      }
    };

    hiveDispatcher.add<TileEvent>('tileSelected', selectionChangeHandler);
    hiveDispatcher.add<TileEvent>('tileDeselected', deselectionChangeHandler);

    return () => {
      hiveDispatcher.remove<TileEvent>('tileSelected', selectionChangeHandler);
      hiveDispatcher.remove<TileEvent>('tileDeselected', deselectionChangeHandler);
      closeConnection().then();
    };
  }, []);

  return (
    <>
      <GameArea players={gameState.players} cells={gameState.cells} playerId={playerId} />
      <Links setShowHelp={setShowHelp} />
      <Rules showHelp={showHelp} setShowHelp={setShowHelp} />
    </>
  );
};

App.displayName = 'App';
export default App;
