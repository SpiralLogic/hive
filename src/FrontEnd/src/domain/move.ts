import {GameId} from "./game-state";
import {HexCoordinates} from './hex-coordinates';
import {TileId} from './tile';

export type Move = {
    tileId: TileId;
    coords: HexCoordinates;
};

export type MoveTile = (gameId: GameId, move: Move) => void;
