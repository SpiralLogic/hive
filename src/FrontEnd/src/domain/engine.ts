import { GameId, GameState } from './game-state';
import { HubConnectionState } from '@microsoft/signalr';
import { Move } from './move';
import { Tile } from './tile';

export type GameStateUpdateHandler = (gameState: GameState) => void;
export type OpponentSelectionHandler = (type: 'select' | 'deselect', tile: Tile) => void;
export type GameConnection = {
  getConnectionState: () => HubConnectionState;
  closeConnection: () => Promise<void>;
  sendSelection: OpponentSelectionHandler;
};
export type GameHandlers = {
  updateGameState: GameStateUpdateHandler;
  opponentSelection?: OpponentSelectionHandler;
};
export type HexEngine = {
  connectGame: (gameId: GameId, handlers: GameHandlers) => GameConnection;
  getNewGame: () => Promise<GameState>;
  moveTile: (gameId: GameId, move: Move) => Promise<GameState>;
  getExistingGame: (gameId: GameId) => Promise<GameState>;
};
