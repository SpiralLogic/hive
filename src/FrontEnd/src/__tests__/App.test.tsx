import { RenderResult, render, screen } from '@testing-library/preact';
import { h } from 'preact';
import { HexEngine, HexServerConnectionFactory } from '../domain/engine';
import { useHiveDispatcher } from '../utilities/dispatcher';
import App from '../components/App';
import { cellMoveEvent, createGameState } from './fixtures/app.fixture';

describe('app tests', () => {
  const connectionFactory: HexServerConnectionFactory = (config) => ({
    connectGame: jest.fn().mockResolvedValue(undefined),
    getConnectionState: jest.fn(),
    closeConnection: jest.fn().mockResolvedValue(undefined),
    sendSelection: jest.fn().mockResolvedValue(undefined),
  });

  const renderApp = (url: string): [RenderResult, HexEngine] => {
    const gameState = createGameState(1);
    const gameAfterMove = createGameState(2);
    const engine: HexEngine = {
      initialGame: Promise.resolve(gameState),
      playerId: 0,
      move: jest.fn().mockResolvedValue(gameAfterMove),
      getNewGame: jest.fn().mockResolvedValue(gameState),
      getExistingGame: jest.fn().mockResolvedValue(gameState),
    };

    global.window.history.replaceState({}, global.document.title, url);

    const app = render(<App engine={engine} connectionFactory={connectionFactory} />);

    return [app, engine];
  };

  it('shows loading', () => {
    renderApp('/');
    expect(screen.getByText(/loading/)).toBeInTheDocument();
  });

  it('shows game when loaded', async () => {
    const [app, engine] = renderApp('/');
    await engine.initialGame;
    app.rerender(<App engine={engine} connectionFactory={connectionFactory} />);
    expect(screen.getByTitle('Hive Game Area')).toBeInTheDocument();
  });

  it('moveTile is called on move events', async () => {
    const [app, engine] = renderApp(`/game/33/1`);
    await engine.getNewGame();
    app.rerender(<App engine={engine} connectionFactory={connectionFactory} />);
    useHiveDispatcher().dispatch(cellMoveEvent);

    expect(engine.move).toHaveBeenCalledTimes(1);
  });
});
