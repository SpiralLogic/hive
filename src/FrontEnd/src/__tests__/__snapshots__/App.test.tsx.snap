// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app tests updates game after move 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="hive"
        title="Hive Game Area"
      >
        <aside
          class="players"
        >
          <div
            class="player player1"
            title="Player 2"
          >
            <div
              class="name"
            >
              Player 2
            </div>
            <div
              class="tiles"
            >
              <div
                class="player1 hex tile creature"
                draggable="true"
                tabindex="0"
                title="ant"
              >
                <svg
                  viewBox="0 0 100 100"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <use
                    href="#hex"
                  />
                  <use
                    class="creature ant"
                    href="#ant"
                  />
                </svg>
              </div>
            </div>
          </div>
        </aside>
        <main>
          <div
            class="links"
          >
            <a
              href="http://localhost/game/33/0"
              name="Share game to opponent"
              title="Share"
            >
              <svg
                viewBox="0 0 100 100"
                xmlns="http://www.w3.org/2000/svg"
              >
                <use
                  href="#share"
                />
              </svg>
            </a>
            <a
              href="/"
              name="New game!"
              title="New Game"
            >
              <svg
                viewBox="0 0 100 100"
                xmlns="http://www.w3.org/2000/svg"
              >
                <use
                  href="#new"
                />
              </svg>
            </a>
            <a
              href="#"
              name="Show rules"
              title="Rules"
            >
              ?
            </a>
            <a
              class="github"
              href="https://github.com/SpiralLogic/hive"
              title="Source code"
            >
              <svg
                viewBox="0 0 100 100"
                xmlns="http://www.w3.org/2000/svg"
              >
                <use
                  href="#github"
                />
              </svg>
            </a>
          </div>
          <div
            class="left"
            role="grid"
          >
            <div
              role="none"
            >
              <div
                class="hide"
                role="none"
              >
                <svg
                  viewBox="0 0 100 100"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <use
                    href="#hex"
                  />
                </svg>
              </div>
            </div>
            <div
              role="row"
            >
              <div
                role="cell"
              >
                <svg
                  viewBox="0 0 100 100"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <use
                    href="#hex"
                  />
                </svg>
                <div
                  class="player1 hex tile creature"
                  draggable="true"
                  tabindex="0"
                  title="ant"
                >
                  <svg
                    viewBox="0 0 100 100"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <use
                      href="#hex"
                    />
                    <use
                      class="creature ant"
                      href="#ant"
                    />
                  </svg>
                </div>
              </div>
            </div>
            <div
              role="none"
            >
              <div
                class="hide"
                role="none"
              >
                <svg
                  viewBox="0 0 100 100"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <use
                    href="#hex"
                  />
                </svg>
              </div>
            </div>
          </div>
        </main>
        
        
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="hive"
      title="Hive Game Area"
    >
      <aside
        class="players"
      >
        <div
          class="player player1"
          title="Player 2"
        >
          <div
            class="name"
          >
            Player 2
          </div>
          <div
            class="tiles"
          >
            <div
              class="player1 hex tile creature"
              draggable="true"
              tabindex="0"
              title="ant"
            >
              <svg
                viewBox="0 0 100 100"
                xmlns="http://www.w3.org/2000/svg"
              >
                <use
                  href="#hex"
                />
                <use
                  class="creature ant"
                  href="#ant"
                />
              </svg>
            </div>
          </div>
        </div>
      </aside>
      <main>
        <div
          class="links"
        >
          <a
            href="http://localhost/game/33/0"
            name="Share game to opponent"
            title="Share"
          >
            <svg
              viewBox="0 0 100 100"
              xmlns="http://www.w3.org/2000/svg"
            >
              <use
                href="#share"
              />
            </svg>
          </a>
          <a
            href="/"
            name="New game!"
            title="New Game"
          >
            <svg
              viewBox="0 0 100 100"
              xmlns="http://www.w3.org/2000/svg"
            >
              <use
                href="#new"
              />
            </svg>
          </a>
          <a
            href="#"
            name="Show rules"
            title="Rules"
          >
            ?
          </a>
          <a
            class="github"
            href="https://github.com/SpiralLogic/hive"
            title="Source code"
          >
            <svg
              viewBox="0 0 100 100"
              xmlns="http://www.w3.org/2000/svg"
            >
              <use
                href="#github"
              />
            </svg>
          </a>
        </div>
        <div
          class="left"
          role="grid"
        >
          <div
            role="none"
          >
            <div
              class="hide"
              role="none"
            >
              <svg
                viewBox="0 0 100 100"
                xmlns="http://www.w3.org/2000/svg"
              >
                <use
                  href="#hex"
                />
              </svg>
            </div>
          </div>
          <div
            role="row"
          >
            <div
              role="cell"
            >
              <svg
                viewBox="0 0 100 100"
                xmlns="http://www.w3.org/2000/svg"
              >
                <use
                  href="#hex"
                />
              </svg>
              <div
                class="player1 hex tile creature"
                draggable="true"
                tabindex="0"
                title="ant"
              >
                <svg
                  viewBox="0 0 100 100"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <use
                    href="#hex"
                  />
                  <use
                    class="creature ant"
                    href="#ant"
                  />
                </svg>
              </div>
            </div>
          </div>
          <div
            role="none"
          >
            <div
              class="hide"
              role="none"
            >
              <svg
                viewBox="0 0 100 100"
                xmlns="http://www.w3.org/2000/svg"
              >
                <use
                  href="#hex"
                />
              </svg>
            </div>
          </div>
        </div>
      </main>
      
      
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
